WIP - Fix part of #8423: Add a lint check to ensure that there are no blank lines below the function definition

AthulJoshi opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #8423 .
2. This PR does the following: Add a lint check to ensure that there are no blank lines below the function definition 

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.

AthulJoshi committed:
added check

AthulJoshi committed:
completed adding check

AthulJoshi committed:
Merge remote-tracking branch 'upstream/develop' into lint-check-for-blank-space-below-fn-def

AthulJoshi committed:
skip ci added_tests

AthulJoshi committed:
skip ci added docstring as requested by linter

AthulJoshi commented:
I had checked out all the PR's and the docs of pylint as it was mentioned. And the linter isnt working as expected. I think i got the concept of ```fromlineno``` wrong. Could you help me out? And could you clear my doubt about  that whether the linter is to check blank line below the ```def fn():``` line? Thanks. And I should squash some commits I think.

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1858,6 +1858,37 @@ def process_tokens(self, tokens):
+            node: astroid.nodes.FunctionDef. Node for a function definition
+                in the AST.
+        """
+        line_number = node.fromlineno
+        line_number += 1
+        if linecache.getline(node.root().file, line_number).strip() == b'':
+            self.add_message(
+                'blank-line-below-function-definition', node=node)
+
+
```
Can we use this check to find whether there's a newline in the first line of function definition?

1. functionDefNode.lineno + len(functionDefNode.doc.split('\n')) != functionDefNode.child.lineno

AthulJoshi replied:
Thanks for the suggestion. Ill check and get back.

DubeySandeep reviewed (changes requested):
@AthulJoshi, Thanks for the PR, the changes looks great I've one suggestion PTAL!

DubeySandeep commented:
> could you clear my doubt about that whether the linter is to check blank line below the def fn(): line?


You are correct, I have left a few examples below PTAL:
Correct:
```
def something():
    """Some doc.
    """
    x = 5
def something():
    x = 5
```

Incorrect:
```
def something():
    """Some doc.
    """

    x = 5
def something():


    x = 5
```

AthulJoshi committed:
changed the condition

AthulJoshi commented:
So, I changed the condition to ```functionDefNode.lineno + len(functionDefNode.doc.split('\n')) + 1 != functionDefNode.body[0].lineno``` which I had checked out using astroid.parse() in another file and had worked. But when I use it here, It throws some errors . And this works assuming that all functions has a docstring.

AthulJoshi commented:
I made a mistake putting that 1 in ```functionDefNode.lineno + len(functionDefNode.doc.split('\n')) + 1```  😅. And I am not able to rectify it as many lint errors are popping up which are caused due to this lint checker.

AthulJoshi commented:
Actually, It works. But it gets triggered falsely when the function def statement spreads over more than a line or when there is a comment just below the docstring. So, there could be a condition to check if the line is a comment and increment ```line_number``` till the comments end. But for the other one I have no idea rn.

AthulJoshi commented:
Ive fixed the issue when comments lines are present. But I am unable to push the changes as the lint check I wrote isnt allowing. Is there another way to push?

AthulJoshi commented:
@DubeySandeep Is there a way to push?

AthulJoshi commented:
Sorry If I am asking silly things. I'm just starting out here :). I am learning angular too so i am looking forward to other issues once I get this over.

AthulJoshi commented:
Hi @DubeySandeep I found the solution to my problem. So should I open another PR?

AthulJoshi commented:
I found that using ```functionDefNode.args.args[-1].lineno``` gives the line number in which the function definition ends.

AthulJoshi commented:
@Hudda Should I open another PR? Please help. :D

AthulJoshi committed:
edited my checker to enable pushing

DubeySandeep commented:
Hi @AthulJoshi, Sorry I missed your old comments. Yeah, please push the changes and open the PR for review.

Also, in general, if you want someone to help in your PR then make sure to assign them! (You can either ask someone on gitter or you can write a comment like " Username PTAL" and the oppia bot will assign them to the PR.) 

> But I am unable to push the changes as the lint check I wrote isn't allowing. Is there another way to push?

It's a good indication that your linter function is working correctly but you'll have to fix the all such lint issue and then push the changes.

Let me know if you are facing any issue, I'll be very happy to help you! :)

AthulJoshi committed:
completed the linter

AthulJoshi commented:
I've completed the linter. PTAL @DubeySandeep.

AthulJoshi commented:
@DubeySandeep PTAL

AthulJoshi commented:
I have completed the linter. Please give your suggestions. :D

AthulJoshi commented:
> Unassigning @AthulJoshi since a re-review was requested. @AthulJoshi, please make sure you have addressed all review comments. Thanks!

What was that lol.

AthulJoshi commented:
It's strange that previously the ```functionDefNode.body[0].lineno```  was working properly but now it says index is out of range...

kevintab95 commented:
Closing/reopening to retrigger actions tests.

DubeySandeep commented:
@AthulJoshi, Do you need any help? Let me know if you have any question or if I can help anyway? Thanks! :)

AthulJoshi commented:
I completed the linter and committed it . Just wanted your review 😅. And suggestions. I used the ptal so that you won't miss it sorry for that..

DubeySandeep commented:
@AthulJoshi, The lint test is still failing can you please fix the lint errors?

DubeySandeep commented:
@AthulJoshi, Make sure to remove "WIP" keyword from the PR once it's ready for review!

AthulJoshi committed:
Lint errors resolved

AthulJoshi commented:
@DubeySandeep  There are so many functions which have blank space below them. What should I do now. I am stuck.

AthulJoshi commented:
Im removing all the lint errors that my linter produced and it might take a while.

AthulJoshi committed:
Merge branch 'develop' into lint-check-for-blank-space-below-fn-def

AthulJoshi commented:
@DubeySandeep I removed all the blank lines below function definitions and now when i try to commit, I think the pre-commit hook isn't allowing :( . Attaching the screenshot as I cant figure what happened. 

![Screenshot from 2020-11-25 15-56-](https://user-images.githubusercontent.com/60695862/100216304-20ce5800-2f38-11eb-9009-2e29aae63426.jpg)

AthulJoshi commented:
@DubeySandeep Please check the previous comment.

DubeySandeep commented:
Hi @AthulJoshi, Sorry I missed your comments. In general, we assign the person to the PR whom we are waiting on (or has the next task to do).

I have very less context about the changes you have locally, can you please check that the python_utils.py file have an open_file function and it's in good structure?

AthulJoshi commented:
> Hi @AthulJoshi, Sorry I missed your comments. In general, we assign the person to the PR whom we are waiting on (or has the next task to do).
> 
> I have very less context about the changes you have locally, can you please check that the python_utils.py file have an open_file function and it's in good structure?

Yep, there is an open_file function and it is same as that in the main repository.


![Screenshot from 2020-12-20 11-38-07](https://user-images.githubusercontent.com/60695862/102706525-b1d6eb80-42b8-11eb-915e-499bbb6744b6.png)

AthulJoshi commented:
I cloned again , copy pasted everything and was able to push :D
